protocol-specs/generators/dpos_random_seed_generation/index.js

```diff
@@ -0,0 +1,471 @@
+	const bufferSizes = new Set(buffers.map(buffer => buffer.length));
+	if (bufferSizes.size > 1) {
+		throw new Error('All input for XOR should be same size');
+	}
+	const outputSize = [...bufferSizes][0];
+	const result = Buffer.alloc(outputSize, 0, 'hex');
+
+	for (let i = 0; i < outputSize; i += 1) {
+		// eslint-disable-next-line no-bitwise
+		result[i] = buffers.map(b => b[i]).reduce((a, b) => a ^ b, 0);
+	}
+
+	return result.toString('hex');
+};
+
+const generateSeedOnion = (initialSeed, size) => {
+	const seeds = new Array(size);
+	seeds[0] = hash(initialSeed, 'hex');
+
+	for (let i = 1; i < size; i += 1) {
+		seeds[i] = strippedHash(seeds[i - 1], 'hex');
+	}
+
+	return seeds.reverse();
+};
+
+const generateSeedReveals = ({ delegateList, numberOfBlocks }) => {
+	const seeds = {};
+
+	for (const delegate of delegateList) {
+		const seedsForDelegate = generateSeedOnion(
+			delegate.publicKey,
+			numberOfBlocks,
+		);
+		const counter = 0;
+
+		seeds[delegate.publicKey] = {
+			counter,
+			seeds: seedsForDelegate,
+		};
+	}
+
+	return seeds;
+};
+
+const generateBlocks = ({ startHeight, numberOfBlocks, delegateList }) => {
+	const seedReveals = generateSeedReveals({ delegateList, numberOfBlocks });
+	const numberOfDelegates = delegateList.length;
+
+	return new Array(numberOfBlocks).fill(0).map((_v, index) => {
+		const height = startHeight + index;
+		const { publicKey } = delegateList[index % numberOfDelegates];
+		const seedReveal =
+			seedReveals[publicKey].seeds[seedReveals[publicKey].counter];
+
+		seedReveals[publicKey].counter += 1;
+
+		return {
+			generatorPublicKey: delegateList[index % numberOfDelegates].publicKey,
+			height,
+			seedReveal,
+		};
+	});
+};
+
+const calcRound = (height, blocksPerRound) =>
+	Math.floor((height - 1) / blocksPerRound) + 1;
+
+const roundInfo = (round, blocksPerRound) => {
+	const firstBlockHeight = round * blocksPerRound - blocksPerRound + 1;
+	const lastBlockHeight = round * blocksPerRound;
+	const middleBlockHeight = Math.floor(
+		(firstBlockHeight + lastBlockHeight) / 2,
+	);
+
+	return {
+		firstBlockHeight,
+		lastBlockHeight,
+		middleBlockHeight,
+	};
+};
+
+const generateRandomSeed = (blocks, blocksPerRound) => {
+	// Middle range of a round to validate
+	const middleThreshold = Math.floor(blocksPerRound / 2);
+
+	const lastBlockHeight = blocks[blocks.length - 1].height;
+
+	// Current round
+	const currentRound = calcRound(lastBlockHeight, blocksPerRound);
+	const {
+		firstBlockHeight: firstBlockHeightOfCurrentRound,
+		middleBlockHeight: middleBlockHeightOfCurrentRound,
+	} = roundInfo(currentRound, blocksPerRound);
+
+	if (lastBlockHeight < middleBlockHeightOfCurrentRound) {
+		throw new Error(
+			`Random seed can't be calculated earlier in a round. Wait till you pass middle of round. Current height ${lastBlockHeight}`,
+		);
+	}
+
+	if (
+		lastBlockHeight >= middleBlockHeightOfCurrentRound &&
+		currentRound === 1
+	) {
+		const randomSeed1 = strippedHash(middleThreshold + 1);
```

Comment:
I think it's better to make the input consistent to be Buffer to make it clear how the number is converted to buffer

Reply:
I felt its consistent to convert the number into buffer in one place which is the function itself. Instead of doing it in multiple places. In the implementation, we can have multiple utility functions if we ant.